### ISSUE TITLE
New Features Discussion

### ISSUE THREAD

**Issue by Cartucho** [Cartucho/mAP#1]
Calculating AP for each class given the PR (Precision Recall) curve.
![tvmonitor](https://user-images.githubusercontent.com/15831541/37559147-e45b3dc4-2a18-11e8-8956-1ccccf83d1c8.jpg)

Calculating the mAP:
![map](https://user-images.githubusercontent.com/15831541/37559152-eed3f002-2a18-11e8-83e0-2da3c898194a.jpg)

**Comment by Cartucho** [comment]
Animation:
![map2](https://user-images.githubusercontent.com/15831541/37559643-6738bcc8-2a21-11e8-8a07-ed836f19c5d9.gif)

**Comment by Cartucho** [comment]
Showing area used to calculate the AP:
![ap_pottedplant](https://user-images.githubusercontent.com/15831541/37722176-3fd0f3e4-2d23-11e8-961c-29001ea2a6b8.png)

**Comment by Cartucho** [comment]
![intersection_over_union_-_visual_equation](https://user-images.githubusercontent.com/15831541/37725175-45b9e1a6-2d2a-11e8-8c15-2fb4d716ca9a.png)

**Comment by andikira** [comment]
Many thanks for your work.
I think it will be great if we have an option to save images with ground-truth box and prediction box in one folder like this image. 

![image](https://user-images.githubusercontent.com/37479922/38145898-bf4d6e64-3475-11e8-8018-00d53e9c1bde.png)

**Comment by Cartucho** [comment]
ok cool, so in the `result/` folder (the one created after running the code) we could have a `result/images/` folder.

We could also add more info to that image, don't you agree? It would be nice to see the confidence of the prediction for example.

What do you think?

**Comment by andikira** [comment]
yeah that must be cool, wait, i have no "result" folder after run the code, it will be good if we can generate txt file showing the mAP result. and how to generate curve and diagram above ?
 
i actually want to contribute in this repo but my programming skill is so noob, sorry for that.

**Comment by Cartucho** [comment]
For the plots you must install matplotlib:

`python -mpip install -U matplotlib`

with that, the `results` folder should show up.

**Comment by Cartucho** [comment]
I will add the .txt file to the results with the terminal output

**Comment by andikira** [comment]
yeah very cool after i install matplotlib i have a curve and diagram in result folder but i got error says   "NameError : name 'xrange' is not defined", if this error have a big matter i'll open new issue.
txt file with some details like details per image or IOU details will be awesome.

**Comment by Cartucho** [comment]
yes, please open a new issue, I will fix it asap

I'm working on the images, I will be adding: 
1) IoU info 
2) confidence info

**Comment by andikira** [comment]
I have tried new feature, it's awesome. I can save the image file with bounding box and prediction box inside it so i can see the error visually. 

another feature :
1. We have a ground truth diagram showing number of object for each class in all images. How about diagram which show number of object predicted for each class, like showing how much sofa predicted, how much tv predicted etc. ?   
2. and the txt result, how about insert more information such as number of object (ground-truth and prediction) for each class, or even the precision-recall table ?

thank you very much for your work and your fast response

**Comment by andikira** [comment]
this is so awesome, can't wait for next feature.

**Comment by Cartucho** [comment]
New feature added: Plot the number of predicted objects per class!

**Comment by andikira** [comment]
The new feature is so cool, 
can we have all the data in diagram or curve in txt files (number of predicted object, number of ground-truth object, precision, recall, etc) ? i just see mAP and AP of each object in txt file.

**Comment by Cartucho** [comment]
How should it be organized? Like this?

1. number of ground-truth objects
2. number of predicted objects
3. AP and precision/recall for each class
4. finally, mAP

Btw, should we use `.txt` or `.json`? `.json` would be easier to parse/use from code and smaller.

**Comment by andikira** [comment]
it's up to you to set the order, i think that order is good enough
actually i don't have any experience for .json file, .txt file will be good for me but if you think .json file is better to organize a lot of data, i think that will be nice to use .json.

**Comment by Cartucho** [comment]
New feature: Added previously discussed info to the file `results/results.txt`

**Comment by andikira** [comment]
Great, I'll try it asap,

**Comment by Cartucho** [comment]
@andikira new feature: Adding values to bars

![image](https://user-images.githubusercontent.com/15831541/38921565-ff5ae370-42ed-11e8-9474-808330c7248c.png)

**Comment by Cartucho** [comment]
And info about true and false predictions:

![image](https://user-images.githubusercontent.com/15831541/38921643-2a9117ee-42ee-11e8-9d2b-1bd233f78c42.png)

**Comment by Cartucho** [comment]
I think I will rotate the plots horizontally to make them easier to read!

**Comment by Cartucho** [comment]
@andikira even better now:

![image](https://user-images.githubusercontent.com/15831541/38933180-366b6fca-4310-11e8-99b9-17ad4b159b86.png)

**Comment by andikira** [comment]
Waw that's so awesome, actually i want to make graph like that, that's why i request to insert detail in txt.
This update will make my life easier 😆    

Thank you @Cartucho 👍  (and ready for new error issue from me haha)

**Comment by Cartucho** [comment]
@JeallyBeans great!

Yeah, could you send the file to me? to.cartucho@gmail.com
If you prefer you can also commit it in a PR.

**Comment by Cartucho** [comment]
@JeallyBeans Added! Thank you!

**Comment by JeallyBeans** [comment]
NP :+1:

**Comment by VinayakaPrabhu** [comment]
Hi @Cartucho 

If there was a way to sort the TP and FP images in results that will be helpful. Thank you.

**Comment by VinayakaPrabhu** [comment]
Hi. Implementing roc curve would be helpful. If anyone could help me in this. Thank you in advance

**Comment by 97andrea97** [comment]
Hey, great work.
If I am not wrong, each point in the PR curve represents a confidence threshold at which the Recall has changed. I think it would be a good idea to add the value of the confidence threshold in the plot (so that one can choose the optimal threshold). Or to output such info in another way. What do you think?
